PP-718: Add fairshare support to the formula

bhroam opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-718](https://pbspro.atlassian.net/browse/PP-718)**
* **[PP-795](https://pbspro.atlassian.net/browse/PP-795)**

#### Problem description
* Enhancement: Add fairshare usage to the job_sort_formula
* PTL bug: PTL doesn't revert fairshare usage to the default between tests

#### Cause / Analysis
* PTL bug: Every cycle the scheduler will write out its view of the fairshare usage.  If any change has been made to the fairshare usage file, it will be overwritten.  PTL was resetting the fairshare usage by removing the fairshare usage file.  This was immediately overwritten on the following cycle.

#### Solution description
* Enhancement: See [EDD](https://pbspro.atlassian.net/wiki/display/PD/PP-718%3A+Add+fairshare+usage+values+to+the+job_sort_formula ) for description on the enhancement.
* PTL Bug: pbsfs now writes out a temporary file in sched_priv when it makes a change.  When the scheduler sees this temporary file, it will reread the fairshare usage and remove the temporary file.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -12165,7 +12163,7 @@ def set_fairshare_usage(self, name=None, usage=None):
 
         :param name: The entity to set the fairshare usage of
-        :type name: str or None
+        :type name: str or PbsUser or None
```
Just a nit: replace PbsUser with ``` :py:class:`~ptl.lib.pbs_testlib.PbsUser` ```

hirenvadalia commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -12518,6 +12525,8 @@ class FairshareNode(object):
     the command ``pbsfs``.
 
+    :param name: Name of fairshare node
+    :type str or None
```
Just a nit: replace this line with ```:type name: str or None```

bhroam commented:
I addressed comments from @hirenvadalia.  Thanks for them.   I also fixed a small bug in the scheduler rereading the usage file.  It wasn't resorting the jobs, so the new usage wouldn't be taken into account that cycle.  I also wrote a pbsfs test.

altair4 commented:
-   what is `"Arbitrary"` usage and why would something arbitrary (i.e.
    *not governed by any fixed rules* to cite one dictionary&rsquo;s definition)
    be useful to track/report?  Perhaps it just needs better (helpfully
    descriptive) name.

-   `pbsfs.c`, lines 235, ff:  no need to use standard I/O to create an
    empty file - use `open(USAGE_TOUCH, O_CREAT | O_RDWR);` (or `creat()`).

-   nit:  `fairshare.c`, line 1236 - &ldquo;it&rsquo;s&rdquo; should be &ldquo;its&rdquo;

-   `fairshare.c`, line 1237:  how can things that have arbitrary values be
    *comparable across the tree*?

-   `fairshare.c`, line 1262:  can&rsquo;t the usage always be reread?  What is this
    really trying to convey?

-   `fairshare.c`, line2 1263, ff:  while the `node` argument is described as
    *the fairshare node to zero*, what we see is `node->usage = 1;`

-   `fifo.c`, line 434:  nit - need white space after `if`

-   `fifo.c`, line 438:  nit:  `job's` should be `jobs'`

-   `fifo.c`, lines 456, ff:  is it worth deciding whether to proceed to do
    the `formula_evaluate()`?  It looks like there are paths through this
    code that don&rsquo;t require it.

bhroam commented:
what is "Arbitrary" usage and why would something arbitrary (i.e.
not governed by any fixed rules to cite one dictionary’s definition)
be useful to track/report? Perhaps it just needs better (helpfully
descriptive) name.

[bmann] I renamed it to usage_factor

pbsfs.c, lines 235, ff: no need to use standard I/O to create an
empty file - use open(USAGE_TOUCH, O_CREAT | O_RDWR); (or creat()).

[bmann] The scheduler has always followed the ANSI standard.  I'd rather leave it as is if you don't mind.

nit: fairshare.c, line 1236 - “it’s” should be “its”

[bmann] Done

fairshare.c, line 1237: how can things that have arbitrary values be
comparable across the tree?

[bmann] Fixed with the rename

fairshare.c, line 1262: can’t the usage always be reread? What is this
really trying to convey?

[bmann] I don't understand this comment.  The block comment doesn't say anything about the readability of the usage file.  It says it will reset the usage so that it can be reread.  That is the point to this function.  It will reset the entire tree back to the original 1 value (there are reasons it's 1).

fairshare.c, line2 1263, ff: while the node argument is described as
the fairshare node to zero, what we see is node->usage = 1;

[bmann] I fixed the comment

fifo.c, line 434: nit - need white space after if

[bmann] fixed

fifo.c, line 438: nit: job's should be jobs'

[bmann] fixed

fifo.c, lines 456, ff: is it worth deciding whether to proceed to do
the formula_evaluate()? It looks like there are paths through this
code that don’t require it.

[bmann] If there is a formula set, it should be always be called on every job in the system.  I don't see the paths you are talking about.  If you look where the block of code came from, you'll see it is called on every job back in query_jobs().  It got moved after where fairshare is updated.  If it didn't move, the fairshare factor in the formula would always be one cycle behind.

altair4 commented:
Regarding the *can&rsquo;t the usage always be reread?* question:

> [bmann] I don&rsquo;t understand this comment. The block comment doesn&rsquo;t say
> anything about the readability of the usage file. It says it will
> reset the usage so that it can be reread. That is the point to this
> function. It will reset the entire tree back to the original 1 value
> (there are reasons it&rsquo;s 1).

I believe what the comment was intended to convey was that the reset
causes the usage value to be updated so that an updated value could
be read.  As currently written, it instead implies that the value can&rsquo;t
be reread unless it&rsquo;s reset.

> [bmann] If there is a formula set, it should be always be called on
> every job in the system. I don&rsquo;t see the paths you are talking about.

What I see is that `formula_evaluate()` is called, then a judgement is made
(with `if (!resresv->can_not_run && [...]`) before referencing the computed
value.  Doesn&rsquo;t this mean that the computed value isn&rsquo;t used?

bhroam commented:
I believe what the comment was intended to convey was that the reset
causes the usage value to be updated so that an updated value could
be read. As currently written, it instead implies that the value can’t
be reread unless it’s reset.

[bmann] I see your point, but I believe the comment is correct.  The reason you need to reset the usage is if any entity is no longer in the usage file, that entity would retain its old value.  The purpose of rereading the usage file is to get the current state of the usage in the file.  So while it is true you could call the read_usage() function without calling reset_usage() first, it would be incorrect to do so.

What I see is that formula_evaluate() is called, then a judgement is made
(with if (!resresv->can_not_run && [...]) before referencing the computed
value. Doesn’t this mean that the computed value isn’t used?

[bmann] I see what you're talking about.  Unfortunately what you are asking would require a change to the external design of the code.  Currently the only way to get the evaluated formula value is from a log message that is printed for every job.  If we suddenly stopped evaluating the formula for jobs we know we don't care about, the log message would not be emitted.  While the change control on this message hasn't been established, I'd view it closer to stable than unstable.  It is literally the only way to view the evaluated formula value.  If we provided another way, I'd be happy with changing it.

altair4 commented:
It might be useful to have a bit of information on `read_usage()` vs.
`reset_usage()` in the function&rsquo;s commentary;  I&rsquo;ll leave that up to
you.

As for the second half, it would seem to be useful long term
to have a better fix for the *only way to get the evaluated
formula value is from a log message* problem - perhaps an RFE?

You can do zero, one, or both of these;  I'd be OK with any of those choices.

bhroam commented on src/include/pbs_share.h:
```diff
@@ -44,7 +44,9 @@
 
-#define FORMULA_FSPERC 		"fair_share_perc"
+#define FORMULA_FSPERC 		"fairshare_perc"
+#define FORMULA_FSPERC_DEP	"fair_share_perc"
```
This isn't used in parse.c.  It's a formula keyword.  Since the formula is opaque to us, so there is no real way to tell if they're using the deprecated keyword or not.

arungrover replied:
You are right, but can we not do a string search for fair_share_perc in scheduler when we read the formula and log the message. I think we can also address this when we use parse job_sort_key in parse.c

bhroam replied:
I looked into this.  It turns out that job_sort_key is broken.  You're supposed to be able to use 'fair_share_perc' in it, but you can't.  If you add fair_share_perc into job_sort_key you are told it is invalid.  Since this is broken to begin with, I'd rather not touch it.  You can't put in preempt_prio either.  Not that preempt_prio does anything anymore.  The scheduler always sorts by preemption priority.  I filed PP-819 for fixing or deprecating the job_sort_key keywords

bhroam commented:
It might be useful to have a bit of information on read_usage() vs.
reset_usage() in the function’s commentary; I’ll leave that up to
you.

[bmann] I updated the block comment saying why you need to reset the usage before rereading it.

As for the second half, it would seem to be useful long term
to have a better fix for the only way to get the evaluated
formula value is from a log message problem - perhaps an RFE?
[bmann] Good idea, I submitted PP-818 for this.

bhroam commented on src/scheduler/fairshare.c:
```diff
@@ -1252,3 +1206,67 @@ reset_temp_usage(group_info *head)
+
+/**
+ * @brief
+ *		recursive helper function to calc_arbitrary_usage()
```
Fixed

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -12209,9 +12208,9 @@ def cmp_fairshare_entities(self, name1=None, name2=None):
+        :type name1: str or :py:class:`~ptl.lib.pbs_testlib.PbsUser` or None
         :param name2: name of second entity to compare
-        :type name1: str or None
+        :type name1: str or :py:class:`~ptl.lib.pbs_testlib.PbsUser` or None
```
Fixed

bhroam commented on src/scheduler/job_info.c:
```diff
@@ -4189,11 +4175,13 @@ formula_evaluate(char *formula, resource_resv *resresv, resource_req *resreq)
 		FORMULA_QUEUE_PRIO, resresv->job->queue->priority,
 		FORMULA_JOB_PRIO, resresv->job->priority,
-		FORMULA_FSPERC, resresv->job->ginfo->percentage,
+		FORMULA_FSPERC, resresv->job->ginfo->tree_percentage,
```
I made sure the new fairshare_factor keyword won't divide by zero.  If it would, I use zero for the value instead.  I added text to the EDD to warn admins to not have entities with zero shares if they use the formula.  SInce the formula is opaque to us, there isn't really any more I can do.  The formula evaluation is in a try/except block, so nothing bad happens to the scheduler.  If a formula throws an exception, zero is used for the evaluated value.

bhroam commented:
Finally, Travis is no longer stuck.

altair4 commented:
One more round(?):

-   `fifo.c`, line 364:  it appears there is no correspoing `fclose(fp);`
-   `fifo.c`, line 366:  why is no check made to see whether the `read_usage()` call
    was successful?
-   `pbsfs.c`, lines 241-242:  if you&rsquo;re not going to check the return value from
    the `fopen()` call, you could simply use `fclose(fopen(USAGE_TOUCH, "w"));`

bhroam commented:
Carl: 
fifo.c, line 364: it appears there is no corresponding fclose(fp);

[bmann] Good catch

fifo.c, line 366: why is no check made to see whether the read_usage() call
was successful?

[bmann] I looked at the function and the only way it failed was if one of the parameters was NULL.  It actually went out of its way to detect errors and just print a message.  I can't think of any good that would come of actually detecting any errors from this function(other than the message).  There are basically two actions we can take.  Continue on and basically reset the usage or bail and stop the scheduler.  If we bail and stop the scheduler, the only recourse the admin will have is to delete the usage file and start over which is basically the same thing.  I just voided the function.

pbsfs.c, lines 241-242: if you’re not going to check the return value from
the fopen() call, you could simply use fclose(fopen(USAGE_TOUCH, "w"));

[bmann] Actually fclose(NULL) will seg fault.  I checked for NULL before I called fclose().

There was a couple of other little things I fixed.  One is that only the leaves (entities) of the tree are supposed to be written out.  It turned out that the unknown group was being written out.  This caused me some problems because in my testing it had some usage even though it was empty.  fairshare groups are supposed to be an accumulation of their children.  I now specifically check to make sure I don't write out the unknown group.

Another issue was when I was calculating the usage_factor.  I need the parent's usage_factor to determine the child's usage_factor.  If there was a tree deeper than 2 levels, it wouldn't calculate correctly.  I now calculate the usage_factor before I recurse.

Let me know what you think

altair4 commented:
| pbsfs.c, lines 241-242: if you’re not going to check the return value from
| the fopen() call, you could simply use fclose(fopen(USAGE_TOUCH, "w"));

> [bmann]
>  Actually fclose(NULL) will seg fault.

That would be the point - by not checking for the return value from fopen(), you were
assuming that it would never return NULL.

Latest round looks OK to me.

arungrover commented:
@bhroam your change looks good. Please squash and rebase for merge.

arungrover commented:
merged
